### PR TITLE
Sample replica connector tolerates f connection failures (v4)

### PR DESCRIPTION
**Pull request by nhoriguchi**
Update #70 with following changes:

- used defer statement for channel close,
- undid using `WaitGroup` for simple synchronization,
- fixed `rootPath` in `sample/vendor/vendor.json`,
- improved dummy `ReplicaConnector` and `core/integration_test.go`.

### CODE REVIEW

File: client/message-handling.go [hunk 1]
```diff
@@ -119,7 +119,7 @@ func makeReplicaConnector(replicaID uint32, connector api.ReplicaConnector) repl
 		if err != nil {
 			return nil, fmt.Errorf("Error getting message stream handler: %s", err)
 		}
-		in, err := streamHandler.HandleMessageStream(out)
+		in := streamHandler.HandleMessageStream(out)
 		if err != nil {
 			return nil, fmt.Errorf("Error establishing connection: %s", err)
 		}
```
Comment (sergefdrv) [inline client/message-handling.go:125]: Could you please also remove the orphaned check for error at line 123?

File: sample/net/dummy/connector/replica-connector.go [hunk 2]
```diff
@@ -19,39 +19,35 @@
...
 
 // ReplicaMessageStreamHandler returns the instance previously
 // assigned to replica ID.
 func (c *ReplicaConnector) ReplicaMessageStreamHandler(replicaID uint32) (api.MessageStreamHandler, error) {
-	replica, ok := c.replicas[replicaID]
-	if !ok {
-		return nil, fmt.Errorf("No message stream handler registered for replica %d", replicaID)
-	}
-	return replica, nil
+	return &messageStreamHandler{replicaID, c}, nil
```
Comment (sergefdrv) [inline sample/net/dummy/connector/replica-connector.go:45]: Now, we need to check if `replicaID < uint32(len(c.replicas))`.

File: sample/net/dummy/connector/replica-connector.go [hunk 3]
```diff
@@ -19,39 +19,35 @@
...
 }
 
 // ConnectReplica assigns a replica instance to replica ID.
 func (c *ReplicaConnector) ConnectReplica(replicaID uint32, replica api.MessageStreamHandler) {
-	c.replicas[replicaID] = replica
+	c.replicas[replicaID] <- replica
```
Comment (sergefdrv) [inline sample/net/dummy/connector/replica-connector.go:50]: This line will only wake a single goroutine created by `messageStreamHandler.HandleMessageStream` and waiting on the channel (or block, if there is no one), whereas the API doesn't forbid calling `MessageStreamHandler.HandleMessageStream` multiple times on the same instance. The replica representation should behave in this case as if it was a real replica instance: establish multiple independent connections. 

Closing a channel, on the other hand, is a non-blocking broadcast signal to all goroutines waiting in receive operation on that channel. This issue should be easy to solve by using pre-allocated slices of replica instances and ready channels. Please look at https://github.com/hyperledger-labs/minbft/pull/70#discussion_r263558017.

**Review by sergefdrv — CHANGES_REQUESTED** [review]
Great job! I think we are nearly finished, just few more comments.

File: client/message-handling.go (hunk 1 above)
Comment (nhoriguchi) [reply client/message-handling.go:125]: Fixed it.

File: sample/net/dummy/connector/replica-connector.go (hunk 2 above)
Comment (nhoriguchi) [reply sample/net/dummy/connector/replica-connector.go:45]: Right, it helps us know what happens for invalid input.

**Pull request closed without merging**